### PR TITLE
[SPARK-27319][SQL] Filter out dir based on PathFilter before listing them

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?

In `InMemoryFileIndex`, we should filter out dir based on PathFilter before listing them. This is useful in the following case: Consider we have the following data:
```
hdfs://name:port/root-dir/timestamp=2019-01-01/***
hdfs://name:port/root-dir/timestamp=2019-01-02/***
....
hdfs://name:port/root-dir/timestamp=2019-03-04/***
```
We can set a PathFilter when the only part of directories' data is needed. Before this patch, we may need to trigger a Spark job or many local threads to listing all the directories. After patch, we can filter out the directories before listing them.

## How was this patch tested?

Added new UT tests.

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Hi @cloud-fan, can you help to review? Thanks a lot.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 1]
```diff
@@ -167,36 +167,39 @@ object InMemoryFileIndex extends Logging {
       hadoopConf: Configuration,
       filter: PathFilter,
       sparkSession: SparkSession): Seq[(Path, Seq[FileStatus])] = {
+    // Filter out the directory before listing it leaf files
+    val filteredPaths = paths.filter(filter.accept(_))
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Just `(filter.accept)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 2]
```diff
@@ -167,36 +167,39 @@ object InMemoryFileIndex extends Logging {
...
 
     // Short-circuits parallel listing when serial listing is likely to be faster.
-    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
-      return paths.map { path =>
+    if (filteredPaths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: My concern is whether the filter is supposed to be applied to the original paths? it seems like it's a filter to be applied to what's inside the paths. I'm not clear that's correct, or if it generally is, whether it would ever have an effect in practice? would you write a filter that probably matched your root dirs but also its contents? They're typically things like filtering for extensions or prefixes of the contents

**Comment by HyukjinKwon** [comment]
Where did this filtering happen before this PR, and how much roughly does it improve the perf?

**Comment by ConeyLiu** [comment]
Hi @srowen @HyukjinKwon, thanks for the review. Before this patch, we need to list all files under directories and then filter them based on the `PatchFilter`. A Spark job will be triggered if the number of directories is very big. 

> how much roughly does it improve the perf?

I didn't measure the roughly perf gain. I found this as the following examples:
I have many data stores in HDFS and in the following format:
```
hdfs://name:port/root-dir/timestamp=2019-01-01/***
hdfs://name:port/root-dir/timestamp=2019-01-02/***
....
hdfs://name:port/root-dir/timestamp=2019-03-04/***
```
And sometimes I just need to read part of directories, so I add a PathFilter in the conf. However, it still need to trigger a Spark job to listing them even the needed directory is just one. That's like what I added in the UT test.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 1 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: will address it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: PathFilter should be applied to the all hdfs input path from the doc. So I think it should be work for the directory as well.

**Comment by cloud-fan** [comment]
The patch LGTM but can you enrich the PR description? When reading the PR description, I have no idea why we need to do it.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by ConeyLiu** [comment]
thanks @cloud-fan, I have updated it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala [hunk 3]
```diff
@@ -354,6 +354,49 @@ class FileIndexSuite extends SharedSQLContext {
...
+      for ((scale, expectedNumPar, dirName) <- Seq((50, 1, "nofiltered"), (50, 1, "nonFiltered"))) {
+        withTempDir { dir =>
+          val subDir = new File(dir, dirName)
+          subDir.mkdirs()
+          (0 until scale).foreach { i =>
+            val file = new File(subDir, s"$i=${i}")
+            file.mkdir()
+          }
+          HiveCatalogMetrics.reset()
+          assert(HiveCatalogMetrics.METRIC_PARALLEL_LISTING_JOB_COUNT.getCount() == 0)
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: Nit: use `===` in asserts

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 4]
```diff
@@ -167,36 +167,43 @@ object InMemoryFileIndex extends Logging {
...
+    } else {
+      paths
+    }
 
     // Short-circuits parallel listing when serial listing is likely to be faster.
-    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
-      return paths.map { path =>
+    if (filteredPaths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
+      return filteredPaths.map { path =>
         (path, listLeafFiles(path, hadoopConf, filter, Some(sparkSession)))
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Rather than change it this way, should `listLeafFiles` just check if `filter` matches `path` initially?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: I see, right, it's really a filter on the full path so it doesn't matter when you check. It won't do anything if you're filtering for `.endsWith(".tmp")` or something and listing `/my/path`. It would change behavior, I guess, if you happened to try to list directory `/my/path.tmp` in this case. Now the whole thing would be filtered out, whereas the intent (I think) and current behavior is to only match files underneath.

That's why I wonder what the use case is here.

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 4 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: this maybe still need to trigger a spark job, because the check is pushed down into the executor side.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: I understand your concern. How about we add some debug log to info the filtered directory or files.  Because the PathFilter should be working on a path, people shouldn't suppose it just works on the name of the current path.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala (hunk 3 above)
Comment (ConeyLiu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: No, the PathFilter is applied on a path, not on the leaf file paths. The path can be file or directory. > Is the use case just setting a filter to apply an operation to a subset of dirs?

just as I described in the pr or in the UT test.

> why not filter them out at the caller level?

Configuration should be more flexible. Otherwise, people may need to filter in the user code. Maybe we can add another option in the datafram reader if we must filter by the user code directly.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Right now, it's only applied to the full path of what is listed by the 'leaf file' methods. I'm not sure what you mean here by leaf file path then. The point is to _also_ apply it to the path of parent dirs earlier, right?

I think the use case could be made clearer. I'm guessing it's something like this:

You want to read a subset of all `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd/xxx.csv` files for example. You only want to read, say, `hdfs://name:port/root-dir/timestamp=2019-MM-dd/xxx.csv` files really. You pass a filter that matches something like `*2019-MM-dd*` and use all the `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd` directories as the data paths. It has to list all the `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd` paths before it can figure out that lots are filtered out and never had to be listed.

If that's the case, that seems a lot easier to filter in user code. Is the use case different or more complex?

I'm trying to weigh the problem with the behavior change against a realistic use case that would benefit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala [hunk 5]
```diff
@@ -95,7 +96,7 @@ class FileIndexSuite extends SharedSQLContext {
       withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
         val fileIndex = new InMemoryFileIndex(spark, Seq(path), Map.empty, None)
         val partitionValues = fileIndex.partitionSpec().partitions.map(_.values)
-        assert(partitionValues.length == 2)
+        assert(partitionValues.length === 2)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: can you revert these unnecessary changes?

Comment (srowen) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: I think this change is fine. It explicitly uses the scalatest macro. I recently learned that will also print a useful message for `==`, though I kind of like using the explicit macros

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Hi @srowen, it's OK to filter the path in user code for me. However, it can trigger a Spark job to filter the subdirectory when the directory need be processed is really large by the current implementation. It would be very slowly in user code.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala (hunk 5 above)
Comment (ConeyLiu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: Hi @cloud-fan, just as @srowen mentioned. `===` should print more message than `==`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 2 above)
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Why would anything need a Spark job here... if you're just taking a set of 100 dirs and filtering them according to a predicate before passing to Spark?

Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: I just mention the benefit of the current implementation, right? The directory can have a list of subdirectories. I still don't think the PathFilter just applied on a leaf file. I can close it if you really think this is useless.

**Comment by srowen** [comment]
Hm, I feel like I am still missing something about the implementation. @adrian-ionescu could I ask you to look at the logic here? I think you implemented a lot of the code in question.

There is some value in filtering out dirs as listLeafFiles / bulkListLeafFiles recurses through the tree, because a filter might try to match intermediate directories and can filter them. I'm worried that a filter like `.endsWith(".tmp")` might match dirs instead of leaf files. But otherwise is this a good optimization?

While a user can filter out top-level dirs they don't actually want to examine, I think there's a decent point here about more complex filters on nested intermediate dirs.

If we're worried about matching intermediate directories for patterns intended to match leaf file paths, maybe it's possible to filter after listing dirs, but check the filter against the dir path plus "/" at  the end if it doesn't already end in "/".

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala (hunk 5 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: That's not always the case (see https://github.com/databricks/scala-style-guide/issues/36). I heard that it can compare arrays properly as well. It would be great if we can document it there.

Comment (ConeyLiu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: thanks @HyukjinKwon, will change back.

**Comment by HyukjinKwon** [comment]
retest this please
